### ISSUE TITLE
Add support for IMAP labels

### ISSUE THREAD

**Issue by mrennings** [k9mail/k-9#769]
Plesse add support for IMAP Labels/flags/tags, just like Mozilla Thunderbird does. Thank you.

**Comment by zaletniy** [comment]
:+1:

**Comment by PanderMusubi** [comment]
What needs to be done is:
* read IMAP labels and show these on a message and in folders
* create, edit and remove mapping of IMAP labels to color, order and more meaningful text
* adding and removing IMAP labels on a message

The order would be used to sort multiple IMAP labels (1st sort), fallback is to order the labels on meaningful text (2nd sort) and fallback for that is to sort them on their original value (3rd sort).

These labels can be more useful than the well known stars, especially when using https://en.wikipedia.org/wiki/Getting_Things_Done

The displaying of labels would use up header space of the message view, but it is up to the user to keep meaningful labels short and to add colors. In that way, only one line in the header would be needed. Displaying in folders could be done by color only, only challenge is what to do when multiple colors are added. Thunderbird gives priority to the most important label. For this order can be used. So if a message has labels "C" and "F" which are respectively mapped to "To Do", order=0, color=#EE88EE and to "Waiting", order=1, color=#FF8800, then on message view both label will be shown, first To Do in purple and then Waiting in orange. In folder view, the message will have purple text color in the subject.

See also https://github.com/k9mail/k-9/pull/558 and https://code.google.com/p/k9mail/issues/detail?id=2532 and https://code.google.com/p/k9mail/issues/detail?id=4101

**Comment by igoralmeida** [comment]
JFYI,

> read IMAP labels and show these on a message and in folders

#558 does that (in the ugliest way possible).

>    create, edit and remove mapping of IMAP labels to color,

Not there yet, but close. I lack the Android UI skills to make this usable, but the backend code is pretty much there already.

>    order and more meaningful text
>    adding and removing IMAP labels on a message

will still take some work after #558.

**Comment by tglman** [comment]
Hi,

Ii there any update on this ? i use tags (or call it labels/flags), i see that a pull request is there, will be really cool have this feature :)

**Comment by PanderMusubi** [comment]
Looking forward to the result. You can ask for help at the UI guys

**Comment by Olifair** [comment]
Is there any news on this improvement ? I have read the discussion in #558 but I don't seem to be able to see any tags in my headers (tags coming from Thunderbird), not in my 1&1 IMAP account, nor in my exchange server IMAP account. Does this need to be enabled to work ? 
I would really appreciate to be able to classify my emails on my phone. And nobody is going to make me change from k9mail to anything else ;-)
Cheers and thanks for all the great developments
Oliver

**Comment by PanderMusubi** [comment]
Added bounty for this one, see https://www.bountysource.com/issues/26376747-add-support-for-imap-labels Hope this helps.

**Comment by ccoenen** [comment]
I actually collected these for another issue, but I thought maybe the links are useful in one place: 

* https://code.google.com/archive/p/k9mail/issues/2532 (2010! Seven years!)
* https://code.google.com/archive/p/k9mail/issues/4101 (lots of discussion)
* https://github.com/k9mail/k-9/pull/146 (PR)
* https://github.com/k9mail/k-9/pull/558 (PR)
* https://github.com/k9mail/k-9/issues/769 (this ticket :D )
* https://github.com/k9mail/k-9/issues/2260

And in one of these I found two bountys:

* https://www.bountysource.com/issues/26376747-add-support-for-imap-labels
* https://www.bountysource.com/issues/817243-feature-suggestion-use-imap-keywords-user-defined-flags-tags-labels

**Comment by cketti** [comment]
I believe this feature needs some design work before it is ready to be implemented.

Here's my list of unordered thoughts on this:
* While this might have been motivated by the IMAP feature to store arbitrary flags I don't see why we need to make this an IMAP-only feature. Other account types could have local-only tags.
* In my opinion tags should be account-specific. Imagine I have a personal account and a shared company account. I don't want the tags used in the company account to show up when working with my personal account and vice versa.
* When tags are supported by the app people want to search/filter for them and maybe even sort by tag.
* People might consider messages with (certain) tags important enough to want to keep them on the device even when they fall outside the sync window.
* For accounts that support server-side tags there needs to be a mechanism to map the apps tags to whatever the server is using.
* We need to figure out what tags to show to the user by default. For local-only tags it's only a question of coming up with a set of default tags. For IMAP we probably want to figure out what custom keywords the user is already using and then try to map that to a list of default tags and hide the rest.
* When an IMAP server doesn't support arbitrary keywords we need to support both local and synced tags.
* How should tags be displayed in the message list? I think changing the backgrounds adds way to much noise. Changing the text color is probably not much better. Maybe add some kind of colored indicator?
* "System flags" and tags are different things to the user. I don't like to conflate them using flags like IMAP does it. But to be able to have different data models (server vs. local) we need a new layer that translates data from the remote store to the model we use inside the app. I believe PR #2607 is a first step towards that (at least for IMAP).
* Currently MessageListFragment more or less directly accesses the database and expects to get all the data it wants to display using a single query. That makes it hard to extract some of the data to different database tables. I believe that needs to be changed before we can implement this feature in a clean way.

**Comment by philipwhiuk** [comment]
> "System flags" and tags are different things to the user. I don't like to conflate them using flags like IMAP does it. 

conflicts with 

> While this might have been motivated by the IMAP feature to store arbitrary flags

and the fact that clients like Thunderbird already support this.

> I don't see why we need to make this an IMAP-only feature.

In reality while I mentioned WebDAV briefly, WebDAV doesn't document any way of storing tags. And POP3 obviously doesn't. So synced tags really are IMAP only. Local tags obviously could be stored on any message. I don't think that's a particular problem - you just don't update custom tags if the account doesn't support fetching/setting custom tags.

> How should tags be displayed in the message list? I think changing the backgrounds adds way to much noise. Changing the text color is probably not much better. Maybe add some kind of colored indicator?

We don't have much space. A small chit maybe, next to the attachment chit on the redesign. But I don't think we should spend too much time on how it looks pre-redesign, only that it's considered in the re-design. Changing the background/text also prohibits multiple tags.

MLF is indeed ugly. It's a view and a controller that enforces behaviour on the data model.

I'm more than a bit grumpy if we really think killing the PR is a good idea though.

**Comment by cketti** [comment]
>> "System flags" and tags are different things to the user. I don't like to conflate them using flags like IMAP does it.
>
> conflicts with
>
>> While this might have been motivated by the IMAP feature to store arbitrary flags
>
> and the fact that clients like Thunderbird already support this.

I don't get that. From the user's perspective e.g. marking a message as read and adding a tag are totally unrelated operations in Thunderbird, too. All I wanted to say was that I don't want tags to be saved as flags in our database just because that's how it works with IMAP. It's an implementation detail that we don't have to copy in our internal data model. But of course we will map "K-9 Mail tags" to IMAP flags/keywords. 

> I'm more than a bit grumpy if we really think killing the PR is a good idea though.

I don't see all the changes I consider necessary for this to be implemented cleanly happening anytime soon. What's the alternative to closing the PR? Merging the code and hoping that someone will clean it up later hasn't worked once in the past.

**Comment by Valodim** [comment]
I do agree that tags should be separate from other flags in our data model. I don't quite see why this can't work in a minimum viable way and sync with imap though? 

At some point we have to work with the resources we actually have, or we set ourselves up for stagnation. It's possible that this is not one of these cases where we a reasonable intermediate feature is achievable, but we should at least make an effort to find out, *especially* if someone is actually doing the work.

**Comment by PanderMusubi** [comment]
I am also in favor of simple independent implementation for IMAP only. First, start to show them, editting, managing names and colors, searching, ordering can be done later.

**Comment by Trogel** [comment]
PR #2668 started as a working prototype. So far it at least helped to spot several design issues that need clarification. In addition, I already learned a lot working on it.  :-)

Personally I don't mind whether the PR is kept alive as work in progress or gets dropped and (some day) succeeded by one or more new PRs. In any case, I am happy to contribute to the design work and the implementation, if there is a chance to make progress. I can spend some hours or a day every now and then.

Regarding cleaning up later, my experience is that it usually never happens. Therefore the fundamental design should be sound before some PR gets merged. At the same time I think we can put some features on the wish list for later.

What is the preferred way to go forward? Continue the discussion here and sketch a high-level overall plan? Raise other, more specific issues or PRs? I can imagine a granularity on the level of "design data model for tags", which might result in some page in the development section on k9mail.github.io, and later "extend local store for tags", then work on IMAP, UI, and so on.

Here are my thoughts on some of the specific issues raised above:

> I don't want tags to be saved as flags in our database just because that's how it works with IMAP.

I agree, we should keep that separate. Since this touches the very base of PR #2668 and its predecessors (8eea618), I also think that starting from scratch makes sense.

> In my opinion tags should be account-specific.

For me one global set of tags would be fine, But I agree that there are use-cases where account-specific tags make sense. And I think, there are also use-cases where one set of tags should be used for more than one account. So, there is a third option - sadly, it is more complex:

Have one or more sets of tags on the global level and link each account to one of these sets.

Changing one of the simpler implementations later into this wouldn't be fun. So, we might better live with the higher complexity from the start. Or is it over the top?

> How should tags be displayed in the message list?

Changing the background color was inspired by Thunderbird. It is exactly what I need, but I did not expect all people to like it. It should be optional, and there should be other options.

In the PR I already considered to optionally add a set of colored tag icons left of the star icon in the message list and message view. The star icon could be made optional also in the message view. And the tag icons could be limited to just one - at the users choice. Just some ideas for a start...

> Changing the background/text also prohibits multiple tags.

Multiple tags can still be assigned to a mail, but this way we can show only one of them in the list. The PR picks the color of the first assigned tag according to the user-defined order, again inspired by Thunderbird. The full list can be seen in the message view.

**Comment by philipwhiuk** [comment]
> Currently MessageListFragment more or less directly accesses the database and expects to get all the data it wants to display using a single query. That makes it hard to extract some of the data to different database tables. I believe that needs to be changed before we can implement this feature in a clean way.

It should be possible to both refactor the tables and keep it as a single query. I'll open a ticket for reviewing/reworking how we organise data in the database.

**Comment by Trogel** [comment]
Are there further opinions regarding the question whether there should be

1. just one global set of tags,
2. one set per account, or
3. multiple sets on the global level with each account being linked to one of them?

I would like to sketch a data model just for the basic sets (or set) of tags, for example on a new page in the development section of the website. This could then serve as a basis for further discussion.

And, of course, please stop me at any time, if you think, I am on the wrong track!

**Comment by severo** [comment]
I imagine the best way would be to have a list of tags per account, possible retrieved from the IMAP server if it implements the functionnality (https://tools.ietf.org/html/rfc3501#section-2.3.2), and if it doesn't, to propose the classical list of the five tags of Thunderbird.

**Comment by ccoenen** [comment]
Just tossing my use case in here for reference. If anyone else has different use cases, it's perhaps also a good idea to comment on that here. If you have the _same_ use-case, i'd suggest just using the reaction-smileys on the specific post. (also: if you think this is a stupid idea, stop me, too ;-) )

So, here it goes:

I am using K-9 with three different accounts. I access those accounts only from K-9 and Thunderbird (with the exception of some webmailer every now and then, but I digress).
I use the standard Thunderbird labels, available via number keys. I do not use them for what they were meant to be, I also don't bother renaming them to my usage. I just want a message to show up as "red" or "green" in my inbox.

To me, having a centralised set of labels would be enough. I do not require @Trogel's option 2 or option 3 (I am not at all against those, I'd just not need them).

**Comment by severo** [comment]
I use the same way as @ccoenen

**Comment by Trogel** [comment]
Thanks for your feedback! Since there is demand for both options 1 and 2, I'd go with option 3 that provides most flexibility.

PR k9mail/k9mail.github.io#57 starts off a design document for message tags as a basis for further discussion. Looking forward to your feedback in the PR!

**Comment by wokawoka** [comment]
Any news on IMAP labels implementation? 
It would be really a great feature.
Since I remember of having seen screenshots documenting the development of this function here on github (or on bountysource?)  some months ago, are there by chance experimental versions with this feature? 
Thanks guys

**Comment by Trogel** [comment]
I am sorry, but I simply ran out of time. Unfortunately I cannot say when the situation will change.

#2668 won't be merged (which is fine for me) and k9mail/k9mail.github.io#57 now looks somewhat over the top to me. Therefore and also since I cannot work on them in the foreseeable future, I will close both pull requests.

Thanks for your great work on k9! It is still my favourite mail client on Android. Just the IMAP keywords are missing...  ;-)

@wokawoka: There are a few screenshots in #2668, but there are no experimental versions of k9 with that feature available, as far as I know.

**Comment by PanderMusubi** [comment]
![badge](https://api.bountysource.com/badge/issue?issue_id=26376747)

Perhaps increasing the bounty will help.

**Comment by h4emp3** [comment]
I too have the exact same use-case as @ccoenen and would very much appreciate this feature being added. And personally I would be totally fine with an absolutely minimal approach, like having K9 just show already existing labels set with another client (e.g. Thunderbird).

**Comment by EnigmaSolved** [comment]
I, too, would love to have this functionality. My use is pretty similar to https://github.com/k9mail/k-9/issues/769#issuecomment-326772984

Also, for handling colors (ie, coloring the background of a message summary in list view) when there are multiple tags: I recommend doing what Thunderbird does and have the highest order tag's color be the one that is used for the background color. I've setup my tags with an understanding of that behavior and it's worked fine for me. And then of course within the message itself display each tag with it's corresponding color. The screenshots from the earlier attempt at all this look right to me: https://github.com/k9mail/k-9/pull/2668#issuecomment-320767174

**Comment by PanderMusubi** [comment]
Any progress on this issue?

**Comment by PanderMusubi** [comment]
Sorry for asking again, any progress? This issue has many likes and with the cooperation with Thunderbird, supporting this will improve similar usage.

**Comment by rezazarchi** [comment]
> * How should tags be displayed in the message list? I think changing the backgrounds adds way to much noise. Changing the text color is probably not much better. Maybe add some kind of colored indicator?

I and My colleague (@mhf1998) designed a demo for message list which may help you to have vision about its design:
![Screenshot_1689062980](https://github.com/thundernest/k-9/assets/8654398/4064e3f0-f809-452f-bb48-3bd1a0575c2f)

**Comment by PanderMusubi** [comment]
![Screenshot from 2023-07-11 13-40-32](https://github.com/thundernest/k-9/assets/198964/d00bb196-f587-487d-b6f7-6c8ec40dc260)

If possible, please support the default colors Thunderbird also uses for these five default tags.

**Comment by ccoenen** [comment]
Should be noted that these **tags are localised**, and therefore you might not be able to rely just on the name at hand! A more international approach might be to scan for used tags and make it really simple to assign a color to them.

**Comment by PanderMusubi** [comment]
Can the localization be supported so it works out of the box? The files needed for this are available in Thunderbird's repo.

**Comment by CoachDom** [comment]
What's the current status of this? I currently sort my Zoho mail using tags so it's a bit painful that all emails to all domains are sitting in one mailbox

**Comment by PanderMusubi** [comment]
I could only find these tests:
- `releases-comm-central-master/mail/components/extensions/test/xpcshell/test_ext_messages.js`
- `releases-comm-central-master/mail/base/test/browser/browser_tagsMode.js`

Where can I find the source files that define the default tags, and possibly where they are translated?

**Comment by PanderMusubi** [comment]
Ah, that was a bit of a search. Here are the tag definitions to (preferably) support by default as they are used too by Thunderbird: https://github.com/mozilla/releases-comm-central/blob/master/mail/locales/en-US/chrome/messenger/messenger.properties#L207

Because the Thunderbird in use can be in another language, the translations of tags can be found in these downloads:
- German https://pontoon.mozilla.org/translations/?code=de&slug=thunderbird&part=mail/chrome/messenger/messenger.properties
- French https://pontoon.mozilla.org/translations/?code=fr&slug=thunderbird&part=mail/chrome/messenger/messenger.properties
- Dutch https://pontoon.mozilla.org/translations/?code=nl&slug=thunderbird&part=mail/chrome/messenger/messenger.properties
- etc.

The locals that are available are: https://github.com/mozilla/releases-comm-central/blob/master/mail/locales/shipped-locales these need to be cross checked with the locales supported by K-9 which are in https://github.com/thundernest/k-9/tree/main/app/ui/legacy/src/main/res (`values-*`). Additionally only well-translated languages from https://explore.transifex.com/k-9/k9mail/

If the developer working on supporting IMAP labels finds it useful, I can make a Python script that does all of the above and generate localized names for default Thunderbird IMAP tags. Perhaps that must be a setting in K-9 to use this mapping or not.

**Comment by PanderMusubi** [comment]
@CoachDom does Zoho mail have any default tags it supports? https://www.zoho.com/mail/help/applying-tags.html?zredirect=f&zsrc=langdropdown&lb=de

**Comment by PanderMusubi** [comment]
Here is a script that collects all tag translations https://pastebin.com/vjPyiMjJ (at the moment, it will only write to the command line)

**Comment by bikb** [comment]
Hello, would it make sense to implement it as an plugin / addon?

**Comment by PanderMusubi** [comment]
For me, this should be standard functionality.

**Comment by bikb** [comment]
So is there a Roadmap available? I think Mozilla is a big institution which plans the development. But where is the plan and who can be assigned to this open issue?

**Comment by qftlzxfz** [comment]
> So is there a Roadmap available? I think Mozilla is a big institution which plans the development. But where is the plan and who can be assigned to this open issue?

I propose to discard previous ideas and discussions and just copy Thunderbird desktop functionality to Thunderbird android.

For anyone using tags it is a major handicap of Thunderbird android not supporting labels.

**Comment by schiessle** [comment]
> I propose to discard previous ideas and discussions and just copy Thunderbird desktop functionality to Thunderbird android.

I agree. I use tags/labels in Thunderbird desktop heavily to organize mails, e.g. I tag  mails as "todo", "read later",... I have the list of tags/label than also on the left sidebar above the folder list to easily access them.

It would be great if Thunderbird on Android would show the tags in mails, allow to set/change tags and also have a easy way to access all mails with a certain tag (maybe also integrate them somehow in the left sidebar as some kind of "virtual folders"),

**Comment by MagicFab** [comment]
If and when this is implemented, please consider supporting user-defined labels, i.e. more than the 5 labels set by default in Thunderbird. 

I personally added 5 more to implement Getting Thing Done tagging using labels. This is done manually via the config editor, see:

https://www.grimeton.com/custom-tags-with-thunderbird-stored-on-imap-server/

ps. Wow, [Bounty Source is gone since May 2024](https://www.reddit.com/r/opensource/comments/1d5v1uk/bountysource_is_dead/).

**Comment by Bugsbane** [comment]
I'm curious if this feature is on any kind of roadmap for the future?

Note: I'm NOT asking/demanding it by any date etc. I get that these things are usually done by volunteers, giving their own precious time where they can spare it for the common good. I'd just love to be able to see where it sits in priority relative to other development work happening. It would be great to be able to look at what's already being released and have some sense of where it's at in the queue.

**Comment by EnigmaSolved** [comment]
> I'm curious if this feature is on any kind of roadmap for the future?
> 
> Note: I'm NOT asking/demanding it by any date etc. I get that these things are usually done by volunteers, giving their own precious time where they can spare it for the common good. I'd just love to be able to see where it sits in priority relative to other development work happening. It would be great to be able to look at what's already being released and have some sense of where it's at in the queue.

I would love to know this as well. :)

**Comment by kewisch** [comment]
You can always see our up to date roadmap at https://github.com/orgs/thunderbird/projects/19/views/1 . Unfortunately this isn't anywhere on the roadmap right now. The closest thing might be #8838 where we'll take a look at the message list, and if the design includes labels we'll know where to slot it in. The above approach to put the labels under seems most straightforward.

To increase chances of this landing on the roadmap I'd recommend to post over on [Mozilla Connect](https://connect.mozilla.org/t5/ideas/idb-p/ideas/label-name/thunderbird%20android) which is one of the prime sources I use to determine roadmap items for my team. The more votes this receives the more likely it will bubble up. Would someone be willing to file an idea there and capture/summarize some of the great constructive comments that have been made here?

**Comment by PanderMusubi** [comment]
Nice to hear about Mozilla Connect. Can the add votes from Bugzilla to their count. There is ~15 years or more of votes.

**Comment by schiessle** [comment]
For everyone interested in the feature, this is probably the right place to up-vote it according to the last comments: https://connect.mozilla.org/t5/ideas/tags-on-thunderbird-android/idi-p/86051
